repoparse/analysis.py

```diff
@@ -0,0 +1,93 @@
+import json
+from sklearn.neighbors import KDTree
+from os import walk
+from typing import Dict, List
+
+
+def read_data(base_path: str = "./out", data_file_prefix: str = "repoparse_"):
+    merged_dict = {}
+    for root, dirs, files in walk(base_path):
+        for f in files:
+            if f.startswith(data_file_prefix):
+                with open(f, "r") as f_reader:
+                    repo_name = f[:len(data_file_prefix)].strip(".json")
+                    loaded_dict = json.load(f_reader)
+                    for _, author_dict in loaded_dict.items():
+                        for _, commit_dict in author_dict.items():
+                            # this is a mishap, please don't hit me too hard
+                            patch_set = commit_dict["patch_languages"]
+                            patch_languages = commit_dict["patch_set"]
+                            commit_dict["patch_languages"] = patch_languages
+                            commit_dict["patch_set"] = patch_set
+                    merged_dict[repo_name] = loaded_dict
+
+    return merged_dict
+
+
+def make_trigrams_cnt(identifier: str) -> Dict[str, int]:
+    res = {}
+    for trigram in map(lambda tup: ''.join(tup), zip(*[identifier[i:] for i in range(3)])):
+        if trigram not in res:
+            res[trigram] = 1
+        else:
+            res[trigram] += 1
+    return res
+
+# TODO generics
+def merge_cnts(left_dict, right_dict) -> dict:
+    result = right_dict.copy()
+    for k, v in left_dict:
+        if k in result:
+            result[k] += v
+        else:
+            result[k] = v
+
+
+def proces_data(merged_dict): # The typing of this thing is too cumbersome
```

Comment:
start with https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.dok_matrix.html - than convert to csr_matrix
```python
lang_mat = dok_matrix(n_devs, n_langs)
ident_mat = dok_matrix(n_devs, n_trigrams)
dev2row = {name: row_n}
lang2col
trig2col
for dev in merged_dict:
   dev_langs = merged_dict[dev]["lang"]
   dev_trig = merged_dict[dev]["trig"]
   for lang in dev_langs:
        lang_mat [dev2row[dev], lang2col[lang]] += dev_langs[lang]
   for trig in dev_trigs:
        ident_mat [dev2row[dev], trig2col[lang]] += dev_trigs[lang]
return lang_mat .to_csr(), ident_mat.to_csr()
```